### PR TITLE
Backend Refactor

### PR DESCRIPTION
**Pull request by joe-elliott**
**What this PR does**:
Simplifies the backend Reader/Writer interfaces to reduce backend complexity and consolidate common logic into one spot. 

This refactor was done to support the blocklist index.

### CODE REVIEW

File: tempodb/backend/backend.go [hunk 1]
```diff
@@ -9,20 +9,23 @@ import (
...
-	ErrEmptyBlockID     = fmt.Errorf("empty block id")
+	ErrDoesNotExist  = fmt.Errorf("does not exist")
+	ErrEmptyTenantID = fmt.Errorf("empty tenant id")
+	ErrEmptyBlockID  = fmt.Errorf("empty block id")
 )
 
 // AppendTracker is an empty interface usable by the backend to track a long running append operation
 type AppendTracker interface{}
 
+// jpe comment all of this
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply tempodb/backend/backend.go:20]: I think so. It consolidates the logic across all backends and keeps the "knowledge" of how to query a Tenant() or Block() list in the backend package. I like both of these things.

File: tempodb/backend/internal.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+type KeyPath []string
+
+// RawWriter is a collection of methods to write data to tempodb backends
+type RawWriter interface {
+	// Write is for in memory data.  It is expected that this data will be cached.
+	Write(ctx context.Context, name string, keypath KeyPath, buffer []byte) error
+	// WriteReader is for larger data payloads streamed through an io.Reader.  It is expected this will _not_ be cached.
+	WriteReader(ctx context.Context, name string, keypath KeyPath, data io.Reader, size int64) error
+	// Append starts or continues an Append job. Pass nil to AppendTracker to start a job.
+	Append(ctx context.Context, name string, keypath KeyPath, tracker AppendTracker, buffer []byte) (AppendTracker, error)
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply tempodb/backend/internal.go:30]: Yes, I think that's the goal ultimately, but currently our caching logic is based on which of these two methods are called. Whenever we add a "cache" parameter to the Read/Write methods we can consolidate on one.

File: tempodb/backend/internal.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// RawReader is a collection of methods to read data from tempodb backends
+type RawReader interface {
+	// Read is for reading entire objects from the backend.  It is expected that there will be an attempt to retrieve this from cache
+	Read(ctx context.Context, name string, keypath KeyPath) ([]byte, error)
+	// ReadReader is for streaming entire objects from the backend.  It is expected this will _not_ be cached.
+	ReadReader(ctx context.Context, name string, keypath KeyPath) (io.ReadCloser, int64, error)
+	// ReadRange is for reading parts of large objects from the backend.  It is expected this will _not_ be cached.
+	ReadRange(ctx context.Context, name string, keypath KeyPath, offset uint64, buffer []byte) error
```
[the remark this answers is not in the document]
Comment (joe-elliott) [reply tempodb/backend/internal.go:42]: Similar comment to above. We can't consolidate until we add a cache param. As far as ReadRange vs. just Read I prefer the separate methods.

1) it allows us to pass a preallocated buffer on ReadRange
2) it is more clear than passing -1 into offset or something to signal "read all"

File: tempodb/backend/internal.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (annanay25) [reply tempodb/backend/internal.go:30]: Gotcha. Will see if I can add that before the cache PR.

**Review by annanay25 — APPROVED** [review]
This is looking good to me.

Pinging @pedrosaraiva to get a review over the Azure specific bits since they've been most active there.

**Pull request merged**